### ISSUE TITLE
Oberthur AuthentIC 3.2 errors on extended length APDU

### ISSUE THREAD

**Issue by lmiguelmh** [OpenSC/OpenSC#1354]
### Problem Description

This is a eID card from Peru and I am using OpenSC 0.18.0-rc1. When using the next command _in Windows 10 everything is fine_:

```
c:\Program Files\OpenSC Project\OpenSC\tools>pkcs11-tool --module "c:\Program Files\OpenSC Project\OpenSC\pkcs11\opensc-pkcs11.dll" --login --test --slot 0
Logging in to "User PIN (OT AWP)".
Please enter User PIN: C_SeedRandom() and C_GenerateRandom():
  seeding (C_SeedRandom) not supported
  seems to be OK
Digests:
  all 4 digest functions seem to work
  MD5: OK
  SHA-1: OK
  RIPEMD160: OK
Signatures (currently only for RSA)
  testing key 0 (Clave privada de autenticaci├│n)
  all 4 signature functions seem to work
  testing signature mechanisms:
    RSA-PKCS: OK
Verify (currently only for RSA)
  testing key 0 (Clave privada de autenticaci├│n)
    RSA-PKCS: OK
Unwrap: not implemented
Decryption (currently only for RSA)
  testing key 0 (Clave privada de autenticaci├│n)
    RSA-PKCS: Mechanism not supported
No errors
```

The _problem happens in Windows 7_, the same command fails:

```
C:\Program Files\OpenSC Project\OpenSC\tools>pkcs11-tool --module "c:\Program Files\OpenSC Project\OpenSC\pkcs11\opensc-pkcs11.dll" --login --test --slot 0
Logging in to "User PIN (OT AWP)".
Please enter User PIN: C_SeedRandom() and C_GenerateRandom():
  seeding (C_SeedRandom) not supported
  seems to be OK
Digests:
  all 4 digest functions seem to work
  MD5: OK
  SHA-1: OK
  RIPEMD160: OK
Signatures (currently only for RSA)
  testing key 0 (Clave privada de autenticaci+¦n)
error: PKCS11 function C_SignFinal failed: rv = CKR_DATA_LEN_RANGE (0x21)
Aborting.
```

In Windows 7 the log shows:

```
P:4700; T:616 2018-05-11 10:31:50.266 reader 'Gemplus USB SmartCard Reader 0'
P:4700; T:616 2018-05-11 10:31:50.266 [opensc-pkcs11] reader-pcsc.c:284:pcsc_transmit: 
Outgoing APDU (44 bytes):
00 2A 9E 9A 00 00 23 30 21 30 09 06 05 2B 0E 03 .*....#0!0...+..
02 1A 05 00 04 14 29 B0 E7 87 82 71 64 5F FF B7 ......)....qd_..
EE C7 DB 4A 74 73 A1 C0 0B C1 04 00             ...Jts......
P:4700; T:616 2018-05-11 10:31:50.266 [opensc-pkcs11] reader-pcsc.c:212:pcsc_internal_transmit: called
P:4700; T:616 2018-05-11 10:31:50.282 [opensc-pkcs11] reader-pcsc.c:293:pcsc_transmit: 
Incoming APDU (2 bytes):
67 00 g.
P:4700; T:616 2018-05-11 10:31:50.282 [opensc-pkcs11] apdu.c:390:sc_single_transmit: returning with: 0 (Success)
P:4700; T:616 2018-05-11 10:31:50.282 [opensc-pkcs11] apdu.c:543:sc_transmit: returning with: 0 (Success)
P:4700; T:616 2018-05-11 10:31:50.282 [opensc-pkcs11] card.c:451:sc_unlock: called
P:4700; T:616 2018-05-11 10:31:50.282 Wrong length
P:4700; T:616 2018-05-11 10:31:50.282 [opensc-pkcs11] iso7816.c:956:iso7816_compute_signature: Card returned error: -1206 (Wrong length)
```
From https://gist.github.com/lmiguelmh/83bf101ecfea925d1cc1aa8ea1c5a244#file-opensc-debug-log-L5951

In Windows 10 the log shows:

```
P:2772; T:5776 2018-05-11 11:20:13.797 reader 'Gemplus USB Smart Card Reader 0'
P:2772; T:5776 2018-05-11 11:20:13.797 [opensc-pkcs11] reader-pcsc.c:284:pcsc_transmit: 
Outgoing APDU (41 bytes):
00 2A 9E 9A 23 30 21 30 09 06 05 2B 0E 03 02 1A .*..#0!0...+....
05 00 04 14 29 B0 E7 87 82 71 64 5F FF B7 EE C7 ....)....qd_....
DB 4A 74 73 A1 C0 0B C1 00                      .Jts.....
P:2772; T:5776 2018-05-11 11:20:13.797 [opensc-pkcs11] reader-pcsc.c:212:pcsc_internal_transmit: called
P:2772; T:5776 2018-05-11 11:20:14.704 [opensc-pkcs11] reader-pcsc.c:293:pcsc_transmit: 
Incoming APDU (258 bytes):
...
90 00                                           ..
P:2772; T:5776 2018-05-11 11:20:14.704 [opensc-pkcs11] apdu.c:390:sc_single_transmit: returning with: 0 (Success)
P:2772; T:5776 2018-05-11 11:20:14.704 [opensc-pkcs11] apdu.c:543:sc_transmit: returning with: 0 (Success)
```
From https://gist.github.com/lmiguelmh/61a2564ca2f42dc97531d949a312364b#file-opensc-debug-windows10-log-L6285

The **difference in the logs is in the APDUs**:

Windows 7:
00 2A 9E 9A 00 00 23 30 21 30 09 ...

Windows 10:
00 2A 9E 9A 23 30 21 30 09 ...

### Logs

Here is the log with debug=9 for Windows 7: 
https://gist.github.com/lmiguelmh/83bf101ecfea925d1cc1aa8ea1c5a244

Here is the log with debug=9 for Windows 10: 
https://gist.github.com/lmiguelmh/61a2564ca2f42dc97531d949a312364b

**Comment by dengert** [comment]
The difference is in using extended APDU vs not extended. 00 00 23 is the extended length for 35 bytes, and at the end is the Lc= 04 00 or 1024 bytes vs 00 meaning read 256 bytes at at time and use GET RESPONSE. 

From the log it lines 13-15:
```
 [opensc-pkcs11] reader-pcsc.c:1063:detect_reader_features: called
Requesting reader features ... 
Gemplus USB SmartCard Reader 0:SCardControl failed: 0x00000001
```

In the Windows 10 log, it get further then in line 19:
```
Assuming that the reader supports transceiving short length APDUs only. 
```

The request failed but at a later time in the code. So in reader-pcsc.c code is executed:
```
1178         /* max send/receive sizes: with default values only short APDU supported */
1179         reader->max_send_size = priv->gpriv->force_max_send_size ?
1180                 priv->gpriv->force_max_send_size :
1181                 SC_READER_SHORT_APDU_MAX_SEND_SIZE;
1182         reader->max_recv_size = priv->gpriv->force_max_recv_size ?
1183                 priv->gpriv->force_max_recv_size :
1184                 SC_READER_SHORT_APDU_MAX_RECV_SIZE;
```
It should be moved up so as to always be done, even if Windows 7 PCSC fails early.  



You could look at setting in opensc.conf:
max_send_size = 255;
max_recv_size = 256;

**Comment by frankmorgner** [comment]
For AuthentIC v3.2 it is assumed that the card is capable of performing extended length. Maybe there's a more intelligent way of detecting the specific card's capabilities, but we need a manual from Oberthur for implementing that.

In the meantime, use the setting in `opensc.conf` as suggested above.

**Comment by lmiguelmh** [comment]
I think this card only works with short length APDUs. 

The same problem is happening in Windows 8 and changing the configuration in `opensc.conf` doesn't worked either.

From @dengert [comment](https://gist.github.com/lmiguelmh/83bf101ecfea925d1cc1aa8ea1c5a244#gistcomment-2586562) I saw the driver was the problem. I've updated the driver in Windows 7 and it is working now. Before the name of the device was "Lector de Tarjeta Inteligente Usbccid de Microsoft (WUDF)" (Spanish) and now is "USB SmartCard Reader".

I've downloaded the driver from: 

 - [Lenovo Gemplus GemPC USB Smart Card Reader driver for Windows 7](https://support.lenovo.com/pe/en/solutions/pd001030)

Thanks for all your help!
